### DISCUSSION TITLE
Any way to make screenshot when soft assertion failed?

### DISCUSSION THREAD

**Discussion by borzykin** [assertj/assertj#2489] [Q&A] [answered]
So as simple as that, can't figure out how to take screenshot at the moment of soft assert failure instead of at the end of test, Using testNg btw

**Comment by ascopes** [comment]
@borzykin not sure how much help this will be, but you could potentially use a custom [soft assertion provider](https://github.com/assertj/assertj-core/blob/main/src/main/java/org/assertj/core/api/SoftAssertionsProvider.java) that just delegates to the default provider implementation.

In this provider, you may be able to try-catch any AssertionError and use [java.awt.Robot](https://docs.oracle.com/javase/9/docs/api/java/awt/Robot.html#createScreenCapture-java.awt.Rectangle-) to capture the primary display device to a BufferedImage. 

I have not used TestNG but you _may_ be able to find a way to infer the test name (maybe by walking the stack trace of the thread), to get a meaningful file name for the image to save.

If you didn't need it for the specific assertion, but just the final assertion error that is thrown, you could just have a lambda that wraps your test case and catches an AssertionError, takes the screenshot using robot, and then rethrows the error again.

**Comment by scordio** [comment, marked as answer]
There are two options:
* Call [`setAfterAssertionErrorCollected`][1] on your `SoftAssertions` instance, providing your own [`AfterAssertionErrorCollected`][2] implementation (can be a lambda).
* Extend `AbstractSoftAssertions` and overrride [`onAssertionErrorCollected​`][3].

Both ways allow you to provide your own logic to be executed with each assertion error.

You can find examples about both solutions in the [AssertJ documentation][4].

_Related: https://stackoverflow.com/questions/68510688_


  [1]: https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/DefaultAssertionErrorCollector.html#setAfterAssertionErrorCollected(org.assertj.core.api.AfterAssertionErrorCollected)
  [2]: https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AfterAssertionErrorCollected.html
  [3]: https://www.javadoc.io/doc/org.assertj/assertj-core/latest/org/assertj/core/api/AfterAssertionErrorCollected.html#onAssertionErrorCollected(java.lang.AssertionError)
  [4]: https://assertj.github.io/doc/#assertj-core-soft-assertions-collect-listener
